packages/state-transition/src/util/seed.ts

```diff
@@ -135,6 +138,46 @@ export function computeProposerIndex(
   );
 }
 
+/**
+ * Return the proposer indices for the given `epoch`.
+ * A more generic version of `computeProposers`
+ */
+export function computeProposerIndices(
+  fork: ForkSeq,
+  state: CachedBeaconStateAllForks,
+  epoch: Epoch
+): ValidatorIndex[] {
+  const startSlot = computeStartSlotAtEpoch(epoch);
+  const proposers = [];
+  const epochSeed = getSeed(state, epoch, DOMAIN_BEACON_PROPOSER);
+  // TODO FULU: Compute shuffling if shuffling cache miss is a temporary workaround.
+  // EpochCache needs to cache shuffling for nextEpoch + 1 too.
+  let shuffling: EpochShuffling;
+  try {
+    shuffling = state.epochCtx.getShufflingAtEpoch(epoch);
+  } catch (e) {
+    state.epochCtx.shufflingCache?.logger?.verbose(
+      `Shuffling cache miss for epoch ${epoch}. Current epoch ${state.epochCtx.epoch}, computing shuffling...`,
+      {},
+      e as Error
+    );
+    state.commit();
+    shuffling = computeEpochShuffling(state, getActiveValidatorIndices(state, epoch), epoch);
```

Comment:
@ensi321 @nflaig we already have `nextShufflingActiveIndices` after `beforeProcessEpoch` so only need to get it from epoch transition cache
the idea is only to do `state.validators` loop once per epoch